Zend/Optimizer/zend_optimizer.c

```diff
@@ -73,7 +73,7 @@ int zend_optimizer_eval_unary_op(zval *result, zend_uchar opcode, zval *op1) /*
 	if (unary_op) {
 		if (opcode == ZEND_BW_NOT
 		 && Z_TYPE_P(op1) != IS_LONG
-		 && Z_TYPE_P(op1) != IS_DOUBLE
+		 /* && Z_TYPE_P(op1) != IS_DOUBLE */ // TODO Optimize when float is int compatible
```

Comment:
I've changed this to use zend_unary_op_produces_error() in https://github.com/php/php-src/commit/ba9c0b38cb141767d99e41789bc96c336608262b, so the change can be dropped from here.